homeassistant/components/hyperion/config_flow.py

```diff
@@ -0,0 +1,395 @@
+        await self._cancel_request_token_task()
+        # Start a task in the background requesting a new token. The next step will
+        # wait on the response (which includes the user needing to visit the Hyperion
+        # UI to approve the request for a new token).
+        self._request_token_task = self.hass.async_create_task(
+            self._request_token_task_func(self._auth_id)
+        )
+        return self.async_external_step(
+            step_id="create_token_external", url=self._get_hyperion_url()
+        )
+
+    async def async_step_create_token_external(
+        self, auth_resp: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle completion of the request for a new token."""
+        if client.ResponseOK(auth_resp):
+            token = auth_resp.get(const.KEY_INFO, {}).get(const.KEY_TOKEN)
+            if token:
+                self._data[CONF_TOKEN] = token
+                return self.async_external_step_done(
+                    next_step_id="create_token_success"
+                )
+        return self.async_external_step_done(next_step_id="create_token_fail")
+
+    async def async_step_create_token_success(
+        self, _: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Create an entry after successful token creation."""
+        # Clean-up the request task.
+        await self._cancel_request_token_task()
+
+        # Test the token.
+        login_ok = await self._can_login()
+
+        if login_ok is None:
+            return self.async_abort(reason="connection_error")
+        if not login_ok:
+            return self.async_abort(reason="auth_new_token_not_work_error")
+        return await self.async_step_confirm()
+
+    async def async_step_create_token_fail(
+        self, _: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Show an error on the auth form."""
+        # Clean-up the request task.
+        await self._cancel_request_token_task()
+        return self.async_abort(reason="auth_new_token_not_granted_error")
+
+    async def async_step_confirm(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Get final confirmation before entry creation."""
+        if user_input is None and self._require_confirm:
+            return self.async_show_form(
+                step_id="confirm",
+                description_placeholders={
+                    CONF_HOST: self._data[CONF_HOST],
+                    CONF_PORT: self._data[CONF_PORT],
+                    CONF_ID: self.unique_id,
+                },
+            )
+
+        async with await self._create_client() as hyperion_client:
+            if not hyperion_client:
+                return self.async_abort(reason="connection_error")
+            hyperion_id = await hyperion_client.async_id()
+
+        if not hyperion_id:
+            return self.async_abort(reason="no_id")
+
+        await self.async_set_unique_id(hyperion_id)
```

Comment:
We may want to set `raise_on_progress` parameter to False here if the user starts a flow and there's already a discovery flow in progress for the same device.

Reply:
Good catch.